bzzeth/bzzeth_test.go

```diff
@@ -123,35 +139,94 @@ func TestBzzEthHandshake(t *testing.T) {
 
 // TestBzzBzzHandshake tests that a handshake between two Swarm nodes
 func TestBzzBzzHandshake(t *testing.T) {
+	// redefine isSwarmNodeFunc to force recognise remote peer as swarm node
+	defer func(f func(*Peer) bool) {
+		isSwarmNodeFunc = f
+	}(isSwarmNodeFunc)
+	isSwarmNodeFunc = func(_ *Peer) bool { return true }
+
 	tester, b, teardown, err := newBzzEthTester()
 	if err != nil {
 		t.Fatal(err)
 	}
 	defer teardown()
 
+	node := tester.Nodes[0]
+	err = handshakeExchange(tester, node.ID(), false, true)
+	if err != nil {
+		t.Fatalf("expected no error, got %v", err)
+	}
+
+	// after successful handshake, expect peer added to peer pool
+	var p *Peer
+	for i := 0; i < 10; i++ {
+		p = b.peers.get(node.ID())
+		if p != nil {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if p == nil {
+		t.Fatal("bzzeth peer not added")
+	}
+
+	// after closing the protocol, expect disconnect
+	close(b.quit)
+	err = tester.TestDisconnected(&p2ptest.Disconnect{Peer: node.ID(), Error: errors.New("?")})
+	if err == nil || err.Error() != "timed out waiting for peers to disconnect" {
+		t.Fatal(err)
+	}
+}
+
+// TestBzzBzzHandshakeWithMessage tests that a handshake between two Swarm nodes and message exchange
+// disconnects the peer
+func TestBzzBzzHandshakeWithMessage(t *testing.T) {
 	// redefine isSwarmNodeFunc to force recognise remote peer as swarm node
 	defer func(f func(*Peer) bool) {
 		isSwarmNodeFunc = f
 	}(isSwarmNodeFunc)
 	isSwarmNodeFunc = func(_ *Peer) bool { return true }
 
+	tester, b, teardown, err := newBzzEthTester()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer teardown()
+
 	node := tester.Nodes[0]
 	err = handshakeExchange(tester, node.ID(), false, true)
 	if err != nil {
 		t.Fatalf("expected no error, got %v", err)
 	}
 
-	// after handshake expect protocol to hang, peer not added to pool
-	p := b.peers.get(node.ID())
-	if p != nil {
-		t.Fatal("bzzeth swarm peer incorrectly added")
+	// after successful handshake, expect peer added to peer pool
+	var p *Peer
+	for i := 0; i < 10; i++ {
+		p = b.peers.get(node.ID())
+		if p != nil {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if p == nil {
+		t.Fatal("bzzeth peer not added")
 	}
 
-	// after closing the ptotocall, expect disconnect
-	close(b.quit)
-	err = tester.TestDisconnected(&p2ptest.Disconnect{Peer: node.ID(), Error: errors.New("protocol returned")})
+	// Send a dummy handshake message, wait for sometime and check if peer is dropped
+	err = dummyHandshakeMessage(tester, node.ID())
 	if err != nil {
 		t.Fatal(err)
 	}
-
+	// after successful handshake, expect peer added to peer pool
+	var p1 *Peer
+	for i := 0; i < 10; i++ {
```

Comment:
This code is pasted three times in this file. It would be nice to have it in a function. As this package will be changed in the near future, it is ok to address this later.

Reply:
Done